### PR TITLE
Remove duplicatable duplicate-input check from CheckTransaction

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Benchmark results indicate this saves about 0.5-0.7ms during CheckBlock.

### CODE REVIEW

**Comment by gmaxwell** [comment]
I believe that with this change if a whitelisted peer gives us a transaction with internally duplicated inputs, we will relay it (CheckTransaction will pass, conflict with the mempool will not set DoS.) and then our peers running without this change will ban us (Their CheckTransaction will fail).

**Comment by TheBlueMatt** [comment]
Simplified diff and fixed.

**Comment by gmaxwell** [comment]
utACK +/- comment nit.

File: src/bench/checkblock.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2016 Matt Corallo
+// Unlike the rest of Bitcoin Core, this file is
+// distributed under the Affero General Public License (AGPL v3)
```
Comment (laanwj) [inline src/bench/checkblock.cpp]: Code under this license cannot be accepted, sorry

**Review by laanwj — CHANGES_REQUESTED** [review]

File: src/bench/checkblock.cpp (hunk 1 above)
Comment (luke-jr) [inline src/bench/checkblock.cpp]: Why?

**Comment by TheBlueMatt** [comment]
Eek, license change was a mistake, sorry.

**Comment by TheBlueMatt** [comment]
Fixed the broken license header.

**Comment by sipa** [comment]
A very interesting duplication of the word "duplicate" in the PR title.

File: src/bench/checkblock.cpp (hunk 1 above)
Comment (gmaxwell) [inline src/bench/checkblock.cpp]: lol. I didn't notice that.

**Comment by sipa** [comment]
@laanwj I believe you need to dismiss the requested changes (as they're addressed).

**Review by laanwj — APPROVED** [review]

**Comment by laanwj** [comment]
Approved the requested change.
BTW: `block413567.hex.h` is a 6MB file - I'd prefer generating this on the fly as we do with the other .X.h files.

**Comment by laanwj** [comment]
Can you please squash the top commit of https://github.com/laanwj/bitcoin/tree/2016_11_generate_bench_file into the benchmark commit?

This has the necessary build system change to generate the block.hex file on the fly from the raw data, like with the tests.

@theuni can you have take a look too perhaps?

**Comment by theuni** [comment]
@laanwj looks good to me. Only slight criticism is that we may want to generate .h/.cpp pairs for blocks to ensure that they end up in separate compilation units, but we can always address that later if it becomes an issue.

File: src/main.cpp [hunk 2]
```diff
@@ -1090,13 +1090,15 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state)
...
-    {
-        if (vInOutPoints.count(txin.prevout))
-            return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-duplicate");
-        vInOutPoints.insert(txin.prevout);
+    // Check for duplicate inputs - note that this check is slow so we skip it in CheckBlock
+    if (fCheckDuplicateInputs) {
+        set<COutPoint> vInOutPoints;
+        for (const auto& txin : tx.vin)
+        {
+            if (vInOutPoints.count(txin.prevout))
```
Comment (theuni) [inline src/main.cpp]: Also worth noting (though not really related here), I measure a ~20%-25% time reduction by avoiding doing 2 lookups:
```c++
std::set<COutPoint> vInOutPoints;
for (const auto& txin : tx.vin)
{
    if (!vInOutPoints.insert(txin.prevout).second)
        return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-duplicate");
}
```

**Comment by laanwj** [comment]
> @laanwj looks good to me. Only slight criticism is that we may want to generate .h/.cpp pairs for blocks to ensure that they end up in separate compilation units, but we can always address that later if it becomes an issue.

Fully agree. On the long run, I'd like to forego the conversion to hex and subsequent compilation entirely on build chains that supports this, by embedding the binary directly as mentioned in a comment here https://github.com/bitcoin/bitcoin/issues/6658#issuecomment-144643696. The current way wastes a lot of time and memory while compiling.

**Comment by TheBlueMatt** [comment]
Picked up (and squashed) @laanwj's changes and went ahead and remove the duplicate-lookup for ATMP uses.

**Comment by ripper234** [comment]
So, if I understand correctly, this caused [CVE-2018-17144](https://bitcoincore.org/en/2018/09/20/notice/).

Can we do a root cause analysis / [5 whys](https://en.wikipedia.org/wiki/5_Whys) explaining how this was introduced?

**Comment by thijstriemstra** [comment]
It's scary to see how this code was merged with so much confidence. The fact there wasn't a unit test covering this piece of code makes me feel little better though, a bad unit test would be even worse. I hope btc core will never accept contributions without tests again.

**Comment by dfoderick** [comment]
> 
> Can we do a root cause analysis / [5 whys](https://en.wikipedia.org/wiki/5_Whys) explaining how this was introduced?

Short version.
This pull request added a boolean parameter to CheckTransaction method  
`bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fCheckDuplicateInputs)`

Then in CheckBlock method the flag was passed in as false. Assumption was that transaction would already be validated by other code so there was no need to validate a second time.  
`if (!CheckTransaction(tx, state, false))`

**Comment by ripper234** [comment]
Why this passed code review? (How many people reviewed it?)

**Comment by ripper234** [comment]
?

**Comment by scravy** [comment]
@dfoderick I question that these questions are in the spirit of a 5 whys root-cause analysis. You are supposed to go deeper, not wider as you do.

Anyhow, the PR merged here was a refactoring and performance improvement, and there already was no test that could have been broken by this; which I believe to be the real culprit.

**Comment by dfoderick** [comment]
> Anyhow, the PR merged here was a refactoring and performance improvement, and there already was no test that could have been broken by this; which I believe to be the real culprit.

Correct. The test for this bug did not exist at the time and was only recently added.
https://github.com/bitcoin/bitcoin/commit/9b4a36effcf642f3844c6696b757266686ece11a

There was a bench test to prove the performance gain but no test for correctness. The bug allows a block with an invalid tx to pass block validation.

**Comment by adlai** [comment]
@ripper234 commented [2 days ago](https://github.com/bitcoin/bitcoin/pull/9049#issuecomment-423837854)
> ?

I'm gonna assume your eloquence was directed at me; if you're not the antecedent of that pronoun, or my assumption was asinine, please ignore this message!

@ripper234, if you truly care about "why this passed code review" (See how here the grammar is correct, and there's no need for editorialization? Amazing, this whole language business!), then you should be able to assign a lower bound to "How many people reviewed it" with confidence equal to your mnemonic fidelity. If this message is unclear, please go back to code review, and leave the Github issue brigading to the _professional_ trolls (or perhaps, take your concerns to the mailing list).

**Comment by jnewbery** [comment]
I suggest we lock this PR. For all the drive-by commenters two years after the fact: if you're serious about wanting to improve security and code quality in Bitcoin Core, you can start by:
- testing release candidates
- reviewing open PRs
- contributing code

**Pull request merged**